File: 349. Intersection of Two Arrays/349.js
// 349. Intersection of Two Arrays solution in JavaScript
// https://leetcode.com/problems/intersection-of-two-arrays/description/

/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var intersection = function(nums1, nums2) {
    let result = [];
    let set_nums1 = new Set(nums1);
    let set_nums2 = new Set(nums2);
    let [small_set, big_set] = set_nums1.length < set_nums2.length ? [set_nums1, set_nums2] : [set_nums2, set_nums1];

    small_set.forEach(num => {
        if (big_set.has(num)) result.push(num);
    });

    return result;
};

File: 349. Intersection of Two Arrays/349.go
// 349. Intersection of Two Arrays solution in GoLang
// https://leetcode.com/problems/intersection-of-two-arrays/description/

func intersection(nums1 []int, nums2 []int) []int {
    nums1_set := make(map[int]bool)
    var result []int

    for _, num := range nums1 {
        nums1_set[num] = true;
    }
    
    for _, num := range nums2 {
        if nums1_set[num] == true {
            result = append(result, num)

            nums1_set[num] = false
        }
    }

    return result
}